### PR TITLE
[grafana] Use sidecar for datasources and dashboards

### PR DESCRIPTION
**Pull request by osterman**
## what
* Update to use sidecar pattern
* Provide integration with `kube-prometheus` (collecting metrics / import grafana dashboards)
* Collect metrics for nginx ingress and display them
* Fix Portal
* Fix nginx ingress template usage
* Update kube-prometheus

## why
* Allow applications to ship their own monitoring configurations
* Integrate independent grafana helm release with kube-prometheus release
* Collect metrics from nginx-ingress with kube-prometheus and display them with grafana
* Fix portal oauth2 ingress bug
* Fix custom errors with external auth
* Use latest kube-prometheus

## todo
- [x] demonstrate loading datasources (they are loading, but grafana doesn't display them)
- [x] demonstrate loading dashboards by YAML/JSON and ID

### CODE REVIEW

**Comment by alebabai** [comment]
Here is a working example with datasource ConfigMap
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: grafana-datasources
  labels:
    grafana_datasource: "true"
data:
  datasource.yaml: |-
    apiVersion: 1
    datasources:
    - name: Prometheus
      type: prometheus
      url: http://kube-prometheus-prometheus:9090
      access: proxy
      isDefault: true
```

New configuration will accessed by grafana on startup (pod should be restarted)

**Comment by alebabai** [comment]
Here is a working example with dashboard ConfigMap

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: grafana-dashboards
  labels:
    grafana_dashboard: "true"
data:
  kubernetes-cluster-monitoring.json: |
    {...}
```

New configuration will accessed by grafana on startup (pod should be restarted)

Also, here is a helper script to download dashboards by id and revision (taken form grafana chart).

```bash
#!/bin/bash

ID=$1
REVISION=$2
DATASOURCE=$3

curl -sk \
    --connect-timeout 60 \
    --max-time 60 \
    -H "Accept: application/json" \
    -H "Content-Type: application/json;charset=UTF-8" \
    https://grafana.com/api/dashboards/$ID/revisions/$REVISION/download | sed "s|\"datasource\":[^,]*|\"datasource\": \"${DATASOURCE}\"|g" \
    > $ID-$REVISION.json

```

File: helmfile.d/0400.kube-prometheus.yaml [hunk 1]
```diff
@@ -152,3 +160,68 @@ releases:
...
+#
+# References:
+#   - https://github.com/coreos/prometheus-operator/tree/master/helm/kube-prometheus
+#   - https://github.com/coreos/prometheus-operator/tree/master/contrib/kube-prometheus
+#   - https://prometheus.io/docs/alerting/configuration
+#   - https://medium.com/@timfpark/simple-kubernetes-cluster-monitoring-with-prometheus-and-grafana-dd27edb1641
+#   - https://www.weave.works/blog/monitoring-kubernetes-infrastructure
+#   - https://www.datadoghq.com/blog/how-to-collect-and-graph-kubernetes-metrics
+#
+- name: "grafana-kube-prometheus"
```
Comment (osterman) [inline helmfile.d/0400.kube-prometheus.yaml]: Rename to grafana-dashboards?

**Review by aknysh — DISMISSED** [review]

File: helmfile.d/0320.nginx-ingress.yaml [hunk 2]
```diff
@@ -66,14 +66,16 @@ releases:
         create: {{ env "NGINX_INGRESS_RBAC_ENABLED" | default "false" }}
       serviceAccount:
         create: {{ env "NGINX_INGRESS_RBAC_ENABLED" | default "false" }}
+
 ## Workaround untill https://github.com/kubernetes/ingress-nginx/pull/2989 would be merged
```
Comment (osterman) [inline helmfile.d/0320.nginx-ingress.yaml]: `until`

Comment (goruha) [reply helmfile.d/0320.nginx-ingress.yaml]: Fixed

File: helmfile.d/0400.kube-prometheus.yaml (hunk 1 above)
Comment (goruha) [reply helmfile.d/0400.kube-prometheus.yaml]: Done

**Review by osterman — COMMENTED** [review]
LGTM +2

**Review by goruha — APPROVED** [review]

**Pull request merged**